typing/types.ml

```diff
@@ -415,6 +413,17 @@ type constructor_description =
     cstr_uid: Uid.t;
    }
 
+and variant_data =
+  { vd_consts: int;                     (* Number of constant constructors *)
+    vd_nonconsts: int;                  (* Number of non-const constructors *)
+    vd_unboxed: int;                    (* Number of unboxed constructors *)
+    mutable vd_max_values: (int bound * int bound) option;
+    (* Maximum immediate value and block tags taking into account unboxed
+       constructors (only available after head-shape analysis) *)
```

Comment:
I'm not convinced by the use of `(int bound * int bound)` here:
- I think it makes the type `max_val_of_cstr_descr` harder to understand (than with, say, a record type with named field)
- is the plan to store all types that are only available later in a single tuple? we discuss that we also need to represent something like `vd_unboxed_consts` and `vd_unboxed_nonconsts, will they also go in this same `mutable ... : ... option` field?

I think that these two issues would be improved by going back to a single field per information, all of them `mutable` with option, but with a helper function for each field in Typedecl

```ocaml
(* types.ml *)
  mutable vd_max_const_tag : int option;

(* typedecl.ml *)
val cstr_max_const_tag : Env.t -> Types.constructor_description -> int
```

that takes care of recomputation and caching.

(I suggest defining the function in `Typedecl` rather than `Typedecl_unboxed` because the user cannot really guess that this logic is related to unboxed constructors.)

Implementation sketch (untested):
```ocaml
(* in typedecl.ml *)
val get_head_shape : Env.t -> Path.t -> head_shape
...

let head_shape_bound = function
  | Shape_any -> Unbounded
  | Shape_set li -> Max (List.fold_left max 0 li)

let get_variant_data env cstr_desc =
  let vd = cstr_desc.variants in
  match vd.vd_max_const_tag, vd.vd_max_nonconst_tag with
  | Some _, Some _ -> vd
  | None, _ | _, None ->
    if vd.vd_unboxed = 0 then begin
      vd.vd_max_const_tag <- Max (vd.vd_consts - 1);
      vd.vd_max_nonconst_tag <- Max (vd.vd_nonconsts - 1);
    end else begin
      let type_path = Types.cstr_type_path cstr_desc in
      let shape = get_head_shape env type_path in
      vd.vd_max_const_tag <- head_shape_bound shape.shape_imm;
      vd.vd_max_nonconst_tag <- head_shape_bound shape.shape_blocks;
  end;
  vd

let cstr_max_const_tag env cstr_descr =
  (get_variant_data env cstr_descr).vd_max_const_tag

let cstr_max_nonconst_tag env cstr_descr =
  (get_variant_data env cstr_descr).vd_max_nonconst_tag
```